src/data/sum/order.lean

```diff
@@ -0,0 +1,518 @@
+                    ⟨to_lex (inl c), inl_lt_inl_iff.2 ha, inl_lt_inl_iff.2 hb⟩
+| inl a, inr b, lex.sep _ _ := let ⟨c, h⟩ := no_bot b in
+                    ⟨to_lex (inr c), inl_lt_inr _ _, inr_lt_inr_iff.2 h⟩
+| inr a, inr b, lex.inr h := let ⟨c, ha, hb⟩ := exists_between h in
+                    ⟨to_lex (inr c), inr_lt_inr_iff.2 ha, inr_lt_inr_iff.2 hb⟩
+end⟩
+
+end lex
+end sum
+
+/-! ### Order isomorphisms -/
+
+open order_dual sum
+
+namespace order_iso
+variables {α β γ : Type*} [has_le α] [has_le β] [has_le γ] (a : α) (b : β) (c : γ)
+
+/-- `equiv.sum_comm` promoted to an order isomorphism. -/
+@[simps apply] def sum_comm (α β : Type*) [has_le α] [has_le β] : α ⊕ β ≃o β ⊕ α :=
+{ map_rel_iff' := λ a b, swap_le_swap_iff,
+  ..equiv.sum_comm α β }
+
+@[simp] lemma sum_comm_symm (α β : Type*) [has_le α] [has_le β] :
+  (order_iso.sum_comm α β).symm = order_iso.sum_comm β α := rfl
+
+/-- `equiv.sum_assoc` promoted to an order isomorphism. -/
+def sum_assoc (α β γ : Type*) [has_le α] [has_le β] [has_le γ] : (α ⊕ β) ⊕ γ ≃o α ⊕ β ⊕ γ :=
+{ map_rel_iff' := by { rintro ((a | a) | a) ((b | b) | b); simp },
+  ..equiv.sum_assoc α β γ }
+
+@[simp] lemma sum_assoc_apply_inl_inl : sum_assoc α β γ (inl (inl a)) = inl a := rfl
+@[simp] lemma sum_assoc_apply_inl_inr : sum_assoc α β γ (inl (inr b)) = inr (inl b) := rfl
+@[simp] lemma sum_assoc_apply_inr : sum_assoc α β γ (inr c) = inr (inr c) := rfl
+@[simp] lemma sum_assoc_symm_apply_inl : (sum_assoc α β γ).symm (inl a) = inl (inl a) := rfl
+@[simp] lemma sum_assoc_symm_apply_inr_inl : (sum_assoc α β γ).symm (inr (inl b)) = inl (inr b) :=
+rfl
+@[simp] lemma sum_assoc_symm_apply_inr_inr : (sum_assoc α β γ).symm (inr (inr c)) = inr c := rfl
+
+/-- `order_dual` is distributive over `⊕` up to an order isomorphism. -/
+def sum_dual_distrib (α β : Type*) [has_le α] [has_le β] :
+  order_dual (α ⊕ β) ≃o order_dual α ⊕ order_dual β :=
+{ map_rel_iff' := begin
+  rintro (a | a) (b | b),
+  { change inl (to_dual a) ≤ inl (to_dual b) ↔ to_dual (inl a) ≤ to_dual (inl b),
+    simp only [to_dual_le_to_dual, inl_le_inl_iff] },
+  { exact iff_of_false not_inl_le_inr not_inr_le_inl },
+  { exact iff_of_false not_inr_le_inl not_inl_le_inr },
+  { change inr (to_dual a) ≤ inr (to_dual b) ↔ to_dual (inr a) ≤ to_dual (inr b),
+    simp only [to_dual_le_to_dual, inr_le_inr_iff] }
+end,
+  ..equiv.refl _ }
+
+@[simp] lemma sum_dual_distrib_inl :
+  sum_dual_distrib α β (to_dual (inl a)) = inl (to_dual a) := rfl
+
+@[simp] lemma sum_dual_distrib_inr :
+  sum_dual_distrib α β (to_dual (inr b)) = inr (to_dual b) := rfl
+
+@[simp] lemma sum_dual_distrib_symm_inl :
+  (sum_dual_distrib α β).symm (inl (to_dual a)) = to_dual (inl a) := rfl
+
+@[simp] lemma sum_dual_distrib_symm_inr  :
+  (sum_dual_distrib α β).symm (inr (to_dual b)) = to_dual (inr b) := rfl
+
+@[pattern] abbreviation inlₗ (x : α) : α ⊕ₗ β := to_lex (sum.inl x)
+@[pattern] abbreviation inrₗ (x : β) : α ⊕ₗ β := to_lex (sum.inr x)
```

Comment:
I expect the linter will want these to have docstrings - feel free to change them to `local notation` if you can make that work.